MSC2285: Private read receipts (#2285)
matrix-org/matrix-spec-proposals: proposals/2285-hidden-read-receipts.md

# MSC2285: Private read receipts

Currently users must send read receipts in order to affect their notification
counts, which alerts other people that the user has read their message. For
primarily privacy reasons, it may be desirable to users to not advertise to
others that they've read a message.

## Proposal

This MSC proposes adding a new `receiptType` (see [the receipts
spec](https://spec.matrix.org/v1.3/client-server-api/#receipts)) of
`m.read.private`. This `receiptType` is used when the user wants to affect their
notification count but doesn't want other users to see their read receipt.

To move the user's private read receipt to `$123` the client can make a POST
request to the [`/receipt`
endpoint](https://spec.matrix.org/v1.3/client-server-api/#post_matrixclientv3roomsroomidreceiptreceipttypeeventid).
For example:

```HTTP
POST /_matrix/client/v3/rooms/!a:example.org/receipt/m.read.private/$123
{}
```

The MSC also proposes adding `m.fully_read` as a possible `receiptType` for

> `/receipt`.

Reviewer: I don't understand the justification for this change from the current state of the document?

Reviewer 2: We are essentially making `m.fully_read` into a `receiptType` which it currently is not. This makes the APIs consistent both `/receipt` and `read_markers` now accept it

Reviewer 2: There isn't a better reason than consistency. 

The original discussion about this centred around the fact that we have two endpoints that are doing essentially the same thing but not really... The idea was to first make them consistent to make things less confusing and then deprecate `/receipt` as it doesn't have much use (and possibly come up with a replacement for `/read_markers` that would support sending some content).

Reviewer: Ah, I was missing that context. It would be useful to put that in the MSC as a justification :)

Alternatively, the client can move the user's `m.fully_read` marker and/or
`m.read` receipt at the same time as `m.read.private` by making a POST request
to the [`/read_markers`
endpoint](https://spec.matrix.org/v1.3/client-server-api/#post_matrixclientv3roomsroomidread_markers).
For example:

```HTTP
POST /_matrix/client/r0/rooms/!a:example.org/read_markers
{
    "m.fully_read": "$123",
    "m.read": "$123",
    "m.read.private": "$123"
}
```

Both `m.read` and `m.read.private` clear notifications in the same way. If the
user sent two receipts into a room, the later one should be the one that decides
the notification count.

If the user has `m.read` and `m.read.private` receipts at the same event, the
`m.read` receipt takes precedence - an `m.read.private` receipt sent to an event
that already has `m.read`, shouldn't move the user's receipt up in the timeline

> from the perspective of other users.

Reviewer 3: I think this entire paragraph can be summarized by "If the m.read.private field is not present, it defaults to the value of m.read"

- m.read is the visible read receipt that everyone can see.
- m.read.private is responsible for notification counts

Reviewer 2: > I think this entire paragraph can be summarized by "If the m.read.private field is not present, it defaults to the value of m.read"

I don't think that is exactly the same thing - if there are both, `m.read` takes precedence (that seems to get lost in your version, I think). Also, I think it is preferable to be more explicit here

> * m.read is the visible read receipt that everyone can see.
> * m.read.private is responsible for notification counts

This is not exactly the case: `m.read` is also responsible for notif counts if there is no `m.read.private` event

Reviewer 4: > This is not exactly the case: m.read is also responsible for notif counts if there is no m.read.private event

Which is why the first sentence said it defaults to the value of m.read, if it is not present.

Reviewer 5: I'm also a bit confused by what's trying to be said here. Why does it matter which one "takes precedence" if they both point to the same event? What would it mean to take one rather than the other? I think this is just to clarify that setting m.read.private has no effect on the m.read receipt and so can't make your read receipt go backwards or mark a message as unread - if so, I'd maybe just say that?

Author: I've tried to clarify this in 5c891d036307f91c9fdd1d701c65e4070b14c5af - please raise new comments :)

should be handled in a separate MSC.

Previous iterations of this MSC additionally suggested that having an `m.hidden`
flag on existing read receipts could work, however this feels like assigning too
much responsibility to an existing structure.

## Unstable prefix

While this MSC is not considered stable, implementations should use
`org.matrix.msc2285` as a namespace.

|Stable (post-FCP)|Unstable                         |
|-----------------|---------------------------------|
|`m.read.private` |`org.matrix.msc2285.read.private`|

## Detecting server support

Clients are required to check for server support to ensure they are not sending
read receipts which are not clearing notifications.

If a client has this feature enabled, in the case of the server not supporting
the MSC, the client should either keep sending private read receipts with the
knowledge that notifications will not be clearing or it should warn the user and
start sending public read receipts.

**Once this MSC gets merged and once it becomes a part of a spec version,
clients should update their implementations as fast as possible to accommodate

> the fact that the way of detecting server support will change.**

Reviewer 6: I still don't think this really calls out the fact that a client has no reliable way to tell if a server supports private read receipts.

```suggestion
Clients should check for server support before sending private read receipts:
if the server does not support them, then a private read receipt will not clear
any notifications for the user.

The presence of `org.matrix.msc2285` or `org.matrix.msc2285.stable` in
`unstable_features` is a reliable indication that a server supports private read
receipts; however the converse is not true: their absence does not necessarily 
mean that the server does *not* support private read receipts. In particular,
the server may have been updated to a future spec version which includes
private read receipts, and hence removed the `unstable_features` entry.

Therefore, if a client has this feature enabled, but the server does not advertise
support for this MSC in `unstable_features`, 
the client should either keep sending private read receipts with the
risk that notifications will not be clearing, or it should warn the user and
start sending public read receipts instead.

To mitigate this problem, once this MSC gets merged and once it becomes a part of a spec version,
clients should update their implementations as fast as possible to accommodate
the fact that the way of detecting server support will change.

Reviewer 2: This doesn't seem to take into account that the `versions` field?

Reviewer 6: well, I think it's implicit, but to be explicit we could change the last paragraph to something like...

```
To mitigate this problem, once this MSC gets merged and once it becomes a part of a spec version,
clients should update their implementations to check for the new spec version in the `versions` field.
```

Author: adopted into https://github.com/matrix-org/matrix-spec-proposals/pull/2285/files#r919275805

> ## Detecting server support
> 
> Clients are required to check for server support to ensure they are not sending
> read receipts which are not clearing notifications.
> 
> If a client has this feature enabled, in the case of the server not supporting
> the MSC, the client should either keep sending private read receipts with the
> knowledge that notifications will not be clearing or it should warn the user and
> start sending public read receipts.
> 
> **Once this MSC gets merged and once it becomes a part of a spec version,
> clients should update their implementations as fast as possible to accommodate
> the fact that the way of detecting server support will change.**
> 
> ### While the MSC is unstable
> 
> During this period, to detect server support clients should check for the
> presence of the `org.matrix.msc2285` flag in `unstable_features` on `/versions`.
> Clients are also required to use the unstable prefixes (see [unstable
> prefix](#unstable-prefix)) during this time.
> 
> ### Once the MSC is merged but not in a spec version
> 
> Once this MSC is merged, but is not yet part of the spec, clients should rely on
> the presence of the `org.matrix.msc2285.stable` flag in `unstable_features` to
> determine server support. If the flag is present, clients are required to use
> stable prefixes (see [unstable prefix](#unstable-prefix)).
> 
> ### Once the MSC is in a spec version
> 
> Once this MSC becomes a part of a spec version, clients should rely on the
> presence of the spec version, that supports the MSC, in `versions` on
> `/versions`, to determine support. Servers will also drop the
> `org.matrix.msc2285.stable` flag once the MSC is part of a spec version.

Author: In an effort to adopt @richvdh's [phrasing](https://github.com/matrix-org/matrix-spec-proposals/pull/2285/files#r919239757) and some feedback internal to the SCT on this:

```suggestion
Clients should check for server support before sending private read receipts:
if the server does not support them, then a private read receipt will not clear
any notifications for the user.

The presence of `org.matrix.msc2285` or `org.matrix.msc2285.stable` in
`unstable_features` is a reliable indication that a server supports private read
receipts; however the converse is not true: their absence does not necessarily
mean that the server does *not* support private read receipts. In particular,
the server may have been updated to a future spec version which includes
private read receipts, and hence removed the `unstable_features` entry.

Therefore, if a client has this feature enabled, but the server does not advertise
support for this MSC in `unstable_features`, the client should either keep sending
private read receipts with the risk that notifications will not be clearing, or it
should warn the user and start sending public read receipts instead.

To mitigate this problem, once this MSC gets merged and once it becomes a part of a
spec version, clients should update their implementations as fast as possible to
accommodate the fact that the way of detecting server support will change: clients
will now be looking for that spec version in `/versions`.

### While the MSC is unstable

During this period, to detect server support clients should check for the
presence of the `org.matrix.msc2285` flag in `unstable_features` on `/versions`.
Clients are also required to use the unstable prefixes (see [unstable
prefix](#unstable-prefix)) during this time.

### Once the MSC is merged but not in a spec version

Once this MSC is merged, but is not yet part of the spec, clients should rely on
the presence of the `org.matrix.msc2285.stable` flag in `unstable_features` to
determine server support. If the flag is present, clients are required to use
stable prefixes (see [unstable prefix](#unstable-prefix)).

### Once the MSC is in a spec version

Once this MSC becomes a part of a spec version, clients should rely on the
presence of the spec version, that supports the MSC, in `versions` on
`/versions`, to determine support. Servers are encouraged to keep the
`org.matrix.msc2285.stable` flag around for a reasonable amount of time
to help smooth over the transition for clients. "Reasonable" is intentionally
left as an implementation detail, however the MSC process currently recommends
*at most* 2 months from the date of spec release.
```

I've essentially copy/pasted @richvdh's comment, made some small edits, skipped over the sections for unstable and stable-but-not-really, and finally edited the spec version section (the last one).

Author: reviewed a bit OOB - committing.

# MSC2285: Private read receipts

Currently users must send read receipts in order to affect their notification
counts, which alerts other people that the user has read their message. For
primarily privacy reasons, it may be desirable to users to not advertise to
others that they've read a message.

## Proposal

This MSC proposes adding a new `receiptType` (see [the receipts
spec](https://spec.matrix.org/v1.3/client-server-api/#receipts)) of
`m.read.private`. This `receiptType` is used when the user wants to affect their
notification count but doesn't want other users to see their read receipt.

To move the user's private read receipt to `$123` the client can make a POST
request to the [`/receipt`
endpoint](https://spec.matrix.org/v1.3/client-server-api/#post_matrixclientv3roomsroomidreceiptreceipttypeeventid).
For example:

```HTTP
POST /_matrix/client/v3/rooms/!a:example.org/receipt/m.read.private/$123
{}
```

The MSC also proposes adding `m.fully_read` as a possible `receiptType` for
`/receipt` to make this endpoint consistent with `/read_markers`. (we have two
endpoints that do essentially the same thing, so it would make sense for them to

> be consistent)

Author: Read markers are deliberately not able to control notifications, hence the need for a new receipt type in this case.

> be consistent)

Reviewer 2: You can't - the `m.fully_read` field is required (at least in the current spec - this MSC changes this). Though what @turt2live means by read markers in this case is `m.fully_read` itself, _not_ the `/read_markers` endpoint

> be consistent)

Reviewer 2: Should be updated

> be consistent)

Reviewer 7: I still don't think `m.fully_read` should be added to `/receipt` - see above :)

the notification count.

The receipt that is more "ahead" of the other takes precedence when considering
notifications and a client's rendering of read receipts. This means that given
an ordered set of events A, B, C, and D the public read receipt could be at
point C, private at point A. If the user moves the private receipt from A to B
then the user's notification count is still considered from point C as the public
receipt is further ahead, still. Other users would also see the user's public read
receipt as not having moved. The user can then move the private read receipt
to point D, hopping over the public receipt, to change their notification count.

For clarity, if the public receipt is "fast forwarded" to be at the same position
as the private receipt then the public receipt is broadcast to other users, even
if previously considered private.

Note that like regular read receipts today, neither receipt can cause a backwards
movement: both receipts can only move forwards, but do not have to be ahead of
each other. It's valid to, for example, update a public read receipt which lags
20 messages behind the private one.

> The `m.read` property is now optional for the [`/read_markers`

Reviewer 8: ```suggestion
The `m.fully_read` property is now optional for the [`/read_markers`
```

`m.read` is already optional, and according to [this comment](https://github.com/matrix-org/matrix-spec-proposals/pull/2285#discussion_r920350727) this MSC makes `m.fully_read` optional.

Reviewer 2: Thank you!!

endpoint](https://spec.matrix.org/v1.3/client-server-api/#post_matrixclientv3roomsroomidread_markers)
as sometimes we only want to send `m.read.private`.

Servers MUST NOT send receipts of `receiptType` `m.read.private` to any other
user than the sender. Servers also MUST NOT send receipts of `receiptType`
`m.read.private` to any server over federation.

## Security considerations

Servers could act as if `m.read.private` is the same as `m.read` so the user
must already trust the homeserver to a degree however, and the methods of
notifying the user to the problem are difficult to implement. Users can always
run their own homeservers to ensure it behaves correctly.

## Potential issues

Clients which support read receipts would end up rendering the user's receipt as
jumping down when they send a message. This is no different from how IRC and
similarly bridged users are perceived today.

## Alternatives

It has been suggested to use account data to control this on a per-account
basis. While this might have some benefits, it is much less flexible and would
lead to us inventing a way to store per-account settings in account data which

> should be handled in a separate MSC.

Author: Per other threads: fully read markers are *not* supposed to be used for notifications at all.

> should be handled in a separate MSC.

Reviewer 3: Why not? Because clients already use it differently?

> should be handled in a separate MSC.

Reviewer 7: The concern is that the current text literally says that you'd have to "invent a way to store _per-account settings_ in _account data_", and this is precisely what account data is made for. Maybe the best is to say exactly what you said in the comment: read receipts are _not_ account data and saving some of them in (room) account data does not seem intuitive and consistent - instead of the "would lead to us inventing the way..." part.

> should be handled in a separate MSC.

Reviewer 2: Ah, I see! I've updated the text. @KitsuneRal, does it make more sense now?

> should be handled in a separate MSC.

Reviewer 7: Yes, good enough, thanks.

Previous iterations of this MSC additionally suggested that having an `m.hidden`
flag on existing read receipts could work, however this feels like assigning too

> much responsibility to an existing structure.

Reviewer 7: I can see how this can be an(other) bikeshedding effort, arriving late to top it off, so feel free to dismiss it if you consider it too much of a change, too late.

I think that one more alternative hasn't been explored - using `/read_markers` (and only `/read_markers`, see above) to send `m.read` receipt with the new value shape - instead of a string with the event id, it would take a dict with the event id and `"m.hidden": true/false`. Given that both `m.read` and `m.read.private` are (should be), in essence, a single entity (I can't imagine a reasonable UX where the two would coexist, rather than one preempting another as the user changes the respective setting in their client), reflecting it in a single thing instead of two things is considerably better. The new value shape would make a compatibility barrier: updated clients can check for the feature support as described here, and  send one of two shapes depending on the homeserver support; non-supporting homeservers would reject the private read receipt even if the client sends it, and non-supporting clients would continue to send the old stringy value, which would still be accepted by homeservers.

Again, I understand that I'm coming and trying half of the MSC now - feel free to dismiss. But I think it's something to be considered before the FCP is over.

Author: Read markers and read receipts do coexist in the Element clients - they serve different purposes.

The `m.hidden` sort of approach is what this MSC originally did, and was rejected for doing so. It was too easy for the user's read receipts to accidentally be public.

Reviewer 7: The `m.hidden` as proposed previously did not alter the value shape of the receipt itself though. It was another key-value next to `m.read` and `m.fully_read`. And, sure, it was all too easy to end up exposing private receipts in that case. Having `"m.read": { "m.hidden": true, "event_id": "$evt" }` would not allow the current homeservers to parse it as usual while updated homeservers could take notice and stop broadcasting `m.read`, while still sending the "usual" `m.read` down the line of the client. There's a small problem still that other clients of the same user would not be aware that this `m.read` is private - but given that they are on the receiving end of that very user I'm not sure about the value of this information.

I apologise again - I should have brought this line of reasoning a month ago...

Author: right, I think I can see the advantage/desire to keep the existing `m.read` type but using an intentionally different shape, though that feels like it'd introduce complexity in the form of selective read receipts (by nature of using a dictionary), which is a bit counter to what this MSC aims to achieve.

It'd certainly be a good option for a future/different MSC though, as I can see the benefit of selective read receipts (email bridge to a room, for example).

Reviewer 4: Sending m.read with a "hidden" flag however was much easier for clients as well as servers to implement. Yes, it adds an additional feature to existing read receipts and it does have some edge cases, where on older servers it wouldn't work properly, but I don't think that was a good argument for the added complexity that m.read.private brings. m.read.private was about 3 times as much code to implement in my client than the hidden flag for what basically amounts to the same feature and the ordering issue it introduces has been like half the discussion on this MSC.

Reviewer 9: For me, the potential for private read receipts to leak on an unsupporting server is a deal breaker, so I strongly disagree with not having a separate type. I also don't view `m.read.private` and `m.read` as being the same entity -- one is simply a *marker for myself* of which rooms I've opened since the latest events arrived in them, the other is something I primarily want to *tell others*.

It's unfortunate that marking for myself and informing others was initially conflated under `m.read`. In an ideal world, we'd have made `m.read` private and then had a separate, explicit `m.read.public` on top when wanting to inform other people. But since this isn't how it was initially designed, I think `m.read.private` is the best we can do now while also ensuring strong guarantees about not leaking information.

Reviewer 4: Hm, yes, that sounds reasonable. But in that case I would propose a long term migration strategy to recommend clients to always send the private read receipt (with m.read being optional instead) and eventually maybe deprecate either that m.read affects notifications or that client are allowed to send m.read without m.read.private (unless for appservice usecases or something). That would sidestep having to deal with both for notifications too (in the long term).

Author: Adding such a recommendation would be a bit too far out of scope for this MSC, at least from my perspective. It's not immediately obvious that the whole ecosystem would benefit from such a recommendation: the spec is more in the position to provide the tools people can use rather than dictate flow, for this case.

----

I'm going to close this thread as resolved: so far it's been concluded that future MSCs can do the additional work that gets brought up here, which I think is the correct decision. Rehashing `m.hidden` (or similar) is also not a productive way to use FCP - we're looking for last minute blockers that will prevent the system from working rather than exploring ideas that have already been discussed (and rejected). If folks spot something in the MSC which won't work, please raise that - it's what the FCP is for.

> # MSC2285: Hidden read receipts

Reviewer 10: How about a way to enable them being hidden globally for your entire account? That way you don't have to specify `hidden` every time, which also means that clients who aren't aware of this option won't leak unwanted read receipts.

Reviewer 3: Do you think this can be designed in a second MSC so this one can be accepted sooner?

Reviewer 2: Also, I would personally prefer to have an option to disable the per-room. This would make it impossible

Reviewer 2: @Sorunome, this sounds like mostly the same point as https://github.com/matrix-org/matrix-doc/pull/2285#discussion_r428069098. Is there any reason that this is a separate thread/do you still think should be global?

Currently users must send read receipts in order to affect their notification counts, which
alerts to other people that the user has read their message. For primarily privacy reasons,
it may be desirable to users to not advertise to others that they've read a message.

## Proposal

> When sending a `m.read` receipt, a `m.hidden: true` flag can be included (optional) to tell

Reviewer 11: I always imagined that this would be implemented by having a separate receipt/marker type, say `m.read.private`, which only the user would see.  Having an `m.hidden` flag works, although the body of the `POST /rooms/.../read_markers` request looks a bit weird to me, with some keys meaning "set this marker to this event" and some flags meaning "add this flag to one of the markers".  So it's a bit non-ideal from my POV, but I won't push against it unless other people feel the same way too.

Author: I would tend to agree with using a different event type, or possibly a different API endpoint entirely (as originally proposed somewhere in the history of issues). However, adding a flag does make the implementation for homeservers a lot easier: this is something they can tack on during development or after they've been well-established (like Synapse). 

I definitely agree that the read marker interface is not great, so suggestions for naming there would be great. I went with consistency rather than changing the name for now though.

Reviewer 9: I concur that the `read_markers` body looks weird to me too. I also find the conflation between "read receipts" and "read markers of type `m.read`" a bit odd, if I understood correctly that those concepts are equivalent. Primarily, this is because the endpoint is called `read_markers` instead of mentioning receipts. From my experience with the ecosystem, there is a lot of confusion around this.

It seems to me that there's a need for three different and rather unconnected features:

1. Read markers: a *private* bookmark to indicate the point up to which you've processed the discussion. Since you haven't processed anything after the read marker, you may want to go back to it later and the read marker allows you to do this.
2. Read receipts: *public* indicators of what you've seen last to inform participants that you have seen it.
3. "Hidden" read receipts: purely internal, and hence *private* mechanism to synchronize "unread messages" indicators between a user's devices (while still retaining the ability from 1 as a separate concept).

When stated like this, 3 does not have much in common with 2, except that they are both a type of receipt and that a receipt of type 2 also serves the purpose of a type 3 receipt (but not vice versa).

This makes me wonder why a separate read marker API was added for the RM instead of considering the RM as just another receipt type and using the `/rooms/.../receipt/RECEIPT_TYPE/EVENT_ID` endpoint with a different `RECEIPT_TYPE`. Is it simply to be able to set them both in a single HTTP call?

In any case, I think it would be much preferable if a new `RECEIPT_TYPE` was added for 3, such as the `m.read.private` as suggested by @uhoreg. The `read_markers` endpoint would then also get an analogous key and would continue to serve the purpose of an endpoint for setting receipt-like things in bulk.

This would also solve the issue raised in [an earlier comment](https://github.com/matrix-org/matrix-doc/pull/2285#pullrequestreview-285145506) by making it fail closed: a newer client configured not to send out read receipts would start sending `m.read.private` exclusively, but an older server would simply ignore such a receipt type (or `read_markers` key), which would prevent information leaks.

Reviewer 3: I can confirm that these three different types of read receipts is also how I understood the spec when implementing it in Conduit. 

1. Read markers (an event in account_data, also called fully_read). Can only be set with /read_markers
2. Public read receipts (read receipts sent to other users). Can be set with /receipt or /read_markers
3. Private read receipts (used to synchronize notification counts between devices). Can be set with /receipt or /read_markers. Currently this type is closely coupled with 2, because it uses the same receipt type.

I recommend that we add a "hidden" field to the /read_markers and /receipt requests. If it is true, the request will not touch the public read receipt (2) and only update the private read receipt (3).

This results in the intended functionality with very few required spec/server changes (at least for conduit). A new receipt_type would mean that a lot more data has to be transferred to update the private read receipt

Author: What are people's feelings on this? Judging from the reactions, it's not obviously clear if a new event type is preferred or if the MSC is fine as-is. It's somewhat worth noting that the Synapse implementation passed review, which is an indicator to me that the proposed flag-based design is at least suitable for the problem space.

Author: I'm taking silence to mean that people are okay with seeing this in a future MSC. In particular, I can see the sort of suggestion for server-side handling being a good addition to per-receipt handling which would make for a very easy MSC down the line.

Reviewer 2: We had a [short discussion](https://matrix.to/#/!NasysSDfxKxZBzJJoE:matrix.org/$EqIPOFfzAHSteuujF0-FqSxOKdGOHZp53igGoWNuJmg?via=matrix.org&via=libera.chat&via=envs.net) about this in #matrix-spec with @timokoesters and concluded that we don't really understand why we have `/receipt` and we think that since we have a way to do this in bulk using `/read_markers` we could deprecate `/receipt` to not duplicate functionality. (maybe I've missed something, feel free to add to this)

AFAICT, that would lead to one of three solutions:

+ Separate `receiptType`
+ A new bulk endpoint which would allow for adding `hidden` flag to a specific `receiptType` (`m.read` in this case)
+ A query param flag to `/read_markers` - I don't like this as the flag should only apply to `m.read` not to `m.fully_read`

Reviewer 9: After some more thought, I would strongly prefer that we encode this new kind of private receipt as a separate receipt type, that is `m.read.private`. In general I think the routing decision should be encoded in the type itself so that there is less room for mistakes and information leakage.

The problem of bulk update is a separate one. Setting only a single type and bulk setting several or all types at once should both be supported.

Read markers and receipts are increasingly sounding to me as instances of the same general concept with differing routing decisions. Even the spec [uses the word "marker"](https://spec.matrix.org/v1.2/client-server-api/#receipts) when talking about receipts. So, indeed, why do we have two different APIs for this?

So to resolve the current impasse, I propose that we:

1. Add `m.read.private` as a new receipt type to the `/receipt` API.
2. Add `m.read.private` as an optional JSON key to the `/read_markers` API, similar to `m.read`.
3. Make the `m.fully_read` JSON key in `/read_markers` optional.

This preserves the ability to bulk update all types of receipts. Then, in a future MSC:

1. Duplicate the `POST /_matrix/client/v3/rooms/{roomId}/receipt/{receiptType}/{eventId}` endpoint as `POST /_matrix/client/v3/rooms/{roomId}/read_markers/{receiptType}/{eventId}`. (Alternatively, duplicate `POST /_matrix/client/v3/rooms/{roomId}/read_markers` as `POST /_matrix/client/v3/rooms/{roomId}/receipt`.)
2. Deprecate the `/_matrix/client/v3/rooms/{roomId}/receipt.*` endpoints. (Alternatively, deprecate the `/_matrix/client/v3/rooms/{roomId}/read_markers.*` endpoints.)

If all JSON keys of the dict payload are made optional, point 1 may be unnecessary.

the server to not send it to anyone else. This allows the user to affect their notification
counts without advertising that they've read the message. `m.hidden` defaults to `false`.

For example:
```
POST /_matrix/client/r0/rooms/!a:example.org/receipt/m.read/$123
{
    "m.hidden": true
}
```

The same applies to read markers (which allow you to update your read receipt):
```
POST /_matrix/client/r0/rooms/!a:example.org/read_markers
{
    "m.fully_read": "$123",
    "m.read": "$123",
    "m.hidden": true
}
```

Note that fully read markers are not sent to other users and are local to the user sending them.
Therefore, no changes are required or implied by `m.hidden` for `m.fully_read` - just `m.read`.

Servers processing read receipts MUST NOT send hidden receipts to any other user than the sender.
Servers MUST NOT send hidden receipts over federation to any server.

## Tradeoffs

Clients which support read receipts would end up rendering the user's receipt as jumping down
when they send a message. This is no different from how IRC and similarly bridged users are

> perceived today.

Reviewer 12: For both ease-of-use and ease-of-privacy reasons, this should be an account setting and not be configurable per room.
I think it's bad UX if the read marker jumps. If it doesn't have any purpose, we should hide it. Making "I don't send read receipts" a part of your public profile makes it easy for others to understand why you seem to never read their messages. This would depend on #489, I guess.

Also, i.m.o this should come together with #2558. If you don't want me to know whether you read my message, I at least want to know whether you have received it. (I want to know if something went wrong.)

Reviewer 13: > If you don't want me to know whether you read my message, I at least want to know whether you have received it.

The people who don't want you to know if you read a message very likely also don't want their devices to leak info of when they receive a message.

Reviewer 13: > For both ease-of-use and ease-of-privacy reasons, this should be an account setting and not be configurable per room.

While clients should primarily present it as an account setting, it should be per-room on the account level so clients that want to provide more granular privacy options could do so (it also makes more sense on the per-room level the way this proposal suggests)

Reviewer 12: > > For both ease-of-use and ease-of-privacy reasons, this should be an account setting and not be configurable per room.
> 
> While clients should primarily present it as an account setting, it should be per-room on the protocol level so clients that want to provide more granular privacy options could do so (it also makes more sense on the per-room level the way this proposal suggests)

Okay, I could live with that. Maybe an account setting that specifies the default, with per-room overrides?

in general, the whole proposal seems very hacky to me. We shouldn't have to send weird hidden dummy events.

Reviewer 13: Yeah, defaults with per-room overrides is what Riot does for most settings

> in general, the whole proposal seems very hacky to me. We shouldn't have to send weird hidden dummy events.

They aren't weird hidden dummy events, they're needed to sync read status across your own devices even if you don't want other people to know what you've read

Author: Have people's opinions changed on this? Judging from the reactions, it's not immediately clear if people are demanding this become an account data thing or if a client-centric approach is okay. Personally, I'd prefer to keep it client-centric so I can have a client which is extremely privacy preserving for the cases I need that, but generally have my read receipts be out there.

Reviewer: Why not both? Could we spec an account data boolean which privacy-conscious clients simply choose to ignore?

Author: Both isn't impossible, but it would imply that we can't make a decision. For simplicity's sake taking one approach over the other would be preferred, whatever that may be.

I'm not seeing a lot of opposition to the client-centric approach though.

Reviewer 14: I think both would be better. I personally would like to be able to sync this setting between all my devices. If Element were to add a line in `im.vector.web.settings` to sync it then it might as well just be a specced field in account data so it applies to all clients.

Reviewer 2: To be honest, if this were to be an account data field, I'd like it to be per-room/per-space/per-DMs/per-whatever and that is another problem in itself... So I think we should leave account data out for the moment and figure out something more complex and re-usable (unless there already is something that I am not aware of...)

Author: Per https://github.com/matrix-org/matrix-doc/pull/2285/files#r720985566 it sounds a bit like there's benefit in a server-side setting for this, though I still don't believe that this MSC should be the one to introduce it. It'd certainly be a welcome addition, however.

> # MSC2285: Hidden read receipts

Reviewer 10: Not sure if it has been considered here already, or if it is out-of-scope for this MSC: It might make sense to add an endpoint to change the default behaviour per-user of `m.hidden`, so that read receipts of clients who do not implement this behaviour yet can still be hidden

Currently users must send read receipts in order to affect their notification counts, which
alerts to other people that the user has read their message. For primarily privacy reasons,
it may be desirable to users to not advertise to others that they've read a message.

## Proposal

When sending a `m.read` receipt, a `m.hidden: true` flag can be included (optional) to tell
the server to not send it to anyone else. This allows the user to affect their notification
counts without advertising that they've read the message. `m.hidden` defaults to `false`.

For example:
```
POST /_matrix/client/r0/rooms/!a:example.org/receipt/m.read/$123
{
    "m.hidden": true
}
```

The same applies to read markers (which allow you to update your read receipt):
```
POST /_matrix/client/r0/rooms/!a:example.org/read_markers
{
    "m.fully_read": "$123",
    "m.read": "$123",
    "m.hidden": true
}
```

Note that fully read markers are not sent to other users and are local to the user sending them.
Therefore, no changes are required or implied by `m.hidden` for `m.fully_read` - just `m.read`.

> Servers processing read receipts MUST NOT send hidden receipts to any other user than the sender.
> Servers MUST NOT send hidden receipts over federation to any server.

Reviewer 15: To prevent user from being nagged to disable this privacy related feature other users, other
users should still see a public read receipt pointing to sender's latest event when the sender
chooses to hide own read receipt:
```suggestion
Servers processing read receipts MUST NOT send hidden receipts to any other user than the sender.
Servers MUST NOT send hidden receipts over federation to any server. Instead, they MUST emulate
a public read receipt pointing to the latest room event sent by the said sender for other users and servers,
OR the event pointed by the latest public read receipt by the user, whichever points to a later event.
```

Author: I feel this defeats the purpose of the MSC. Do you have any information on what your suggestion actually means?

Reviewer 2: I think I am overlooking something - how does this defeat the purpose of the MSC?

Reviewer 15: > I feel this defeats the purpose of the MSC.

This change is designed prevent from detection of whether one prefers public or hidden read receipts, without leaking any additional information from the sender. Without this change, other users and servers will easily be able to see that preference, which would cause nagging by room mods and/or server operators (including other servers in the federation that the rooms user joins federate).

Author: This was discussed briefly in #matrix-spec and it seems there is a misunderstanding in what this MSC does. The MSC just blocks propagation of receipts to allow notifications to work properly while not alerting other users to the user's read-up-to status. The MSC does not block implicit read receipts as currently required by the spec.

The suggestion appears to be regurgitating the spec in an awkward way, so I'm going to decline to add it. At first, it appeared as though the suggestion was to send something down with some sort of metadata to make it not work on the receiving side, which is not the case.

## Tradeoffs

Clients which support read receipts would end up rendering the user's receipt as jumping down
when they send a message. This is no different from how IRC and similarly bridged users are
perceived today.

## Security considerations

Servers could ignore the flag without telling the user. The user must already trust the homeserver
to a degree however, and the methods of notifying the user to the problem are difficult to
implement. Users can always run their own homeservers to ensure it behaves correctly.

## Why not X kind of EDUs?

In short: don't send those EDUs. Typing notifications, device messages, etc can all be mitigated
by simply not calling the endpoints. Read receipts have a side effect of causing stuck
notifications for users however, which is why they are solved here.

## Unstable prefix

While this MSC is not considered stable, implementations should use `org.matrix.msc2285` as a namespace

> for identifiers. `m.hidden` becomes `org.matrix.msc2885.hidden` for example.

Reviewer 2: ```suggestion
for identifiers. `m.hidden` becomes `org.matrix.msc2285.hidden` for example.
```

# MSC2285: Hidden read receipts

Currently users must send read receipts in order to affect their notification
counts, which alerts other people that the user has read their message. For
primarily privacy reasons, it may be desirable to users to not advertise to
others that they've read a message.

## Proposal

This MSC proposes adding a new `receiptType` of `m.read.private`. This
`receiptType` is used when the user wants to affect their notification count but
doesn't want other users to see their read receipt.

To move the user's private read receipt to `$123` the client can make a POST
request such as this.

```HTTP
POST /_matrix/client/r0/rooms/!a:example.org/receipt/m.read.private/$123
{}
```

> 

Reviewer 2: ```suggestion
The MSC also proposes adding `m.fully_read` as a possible `receipType` for `/receipt`.

```

To also move the user's `m.fully_read` marker and `m.read` receipt the client
can make a POST request such as this.

```HTTP
POST /_matrix/client/r0/rooms/!a:example.org/read_markers
{
    "m.fully_read": "$123",
    "m.read": "$123",
    "m.read.private": "$123",
}
```

It is assumed that if only an `m.read` receipt is received, the `m.read.private`

> should also be moved.

Reviewer 16: From the Synapse review (matrix-org/synapse#12168) I think we tweaked this logic slightly such that both receipts can co-exist in a room, but the "latest" one is what is actually used. So if you have events `A`, `B`, and `C` with a private receipt to `B`, then send a public receipt to `C` your private receipt doesn't really matter anymore since you have a later one.

I suspect the idea behind this text is actually that `m.read.private` clears notifications in the same way as ` m.read`?

Reviewer 2: Does this sound good?

```suggestion
Both `m.read` and `m.read.private` clear notifications in the same way. If the
user sent two receipts into a room, the later one should be the one that decides
the notification count.

If the user has `m.read` and `m.read.private` receipts at the same event, the
`m.read` receipt takes precedence - an `m.read.private` receipt sent to an event
that already has `m.read`, shouldn't move the user's receipt up in the timeline
from the perspective of other users.
```

Reviewer 16: @SimonBrandner:

> If the user sent two receipts into a room, the latter one should be the one that decides the notification count.

Is "latter" a typo (should it be "later") or does it refer to `m.read.private` taking precedence for notifications?

Reviewer 2: I meant _later_ - thank you for the catch! I've updated my comment

> # MSC2285: Private read receipts

Reviewer 16: My understanding is that as of matrix-org/synapse#12168 and matrix-org/matrix-js-sdk#2221 the MSC matches the implementations.

> # MSC2285: Private read receipts

Reviewer 17: (this comment mostly exists as down votes without comment are frowned upon by some people of the spec core team and previously lead to discussions.)

I personally am in favor of seperating the notification counter from the read marker however I strongly am against having private read receipts.

This has both social and technical reasons.

1. Socially I strongly believe it is good etiquette to have read markers to prevent people from asking multiple times if someone doesn't reply for whatever reason. Surely another solution would be to expect them to write "sorry read but can't reply" but at that point the user is effectively doing what the system was supposed to be solving here. Also most people are too lazy to do this as they expect the system to do it for them.
Also I am aware that email and other email style systems do not do read markers. However I believe they have a different type of audience and different expectations associated with them. Matrix is primarily an _Instant_ Messaging app. Hence imho there should also be instant feedback. While email is an async medium where a reply is expected to only happen in "working hours". Is this a social issue? Maybe. But I don't believe this is a problem matrix spec should be solving which the msc seems to try.

2. On the technical side I believe that we need those read receipts as you may have federation problems which aren't uncommon which you wouldn't detect without read receipts leading to silently lost messages. They are currently they only indicator we have apart from manually running the federation tester to check if there is a broken or half broken fed. This problem could likely be solved by adding a better indicator (aka a HS self health check).

Reviewer 4: According to German (and probably EU) data protection laws, you are required to collect and send the minimal amount of data by default. This is currently not possible. While you can hide your online status, it is broadcasted to everyone everytime you open a room. While this can be nice with your friends, it is not nice when communicating with your employer, doctor or other people you don't consider your friend or who you don't want to know, when you are awake. For some people it is also very uncomfortable to read a message and knowing the other party can see that, but they might not have the energy right now to reply. As such it is important to give people the option in my opinion. Heck, clients can go as far and have different options for groups and DMs.

For context, the upcoming TI messenger specification seems to require hidden read receipts by default too: https://github.com/gematik/api-ti-messenger/tree/main/Vorabveroeffentlichungen/TIM_1.1

Matrix already gives away some metadata, giving control to the user is important imo. You shouldn't be forced to share your personal data. :3

Author: @MTRNord second ping - we'd like to resolve this.

Reviewer 2: (they are possibly on holiday since they're not responding on Matrix either)

Author: Having actually read the comment, I believe this to be resolved and so am closing it. The opinion is noted, however with a "direction of Matrix" hat on, I think we need this MSC for chat and non-chat purposes alike.

Currently users must send read receipts in order to affect their notification
counts, which alerts other people that the user has read their message. For
primarily privacy reasons, it may be desirable to users to not advertise to
others that they've read a message.

## Proposal

> This MSC proposes adding a new `receiptType` of `m.read.private`. This

Reviewer 6: It's really helpful for people reading an MSC if you provide links to relevant existing spec, to provide context. In this case the relevant context appears to be https://spec.matrix.org/v1.3/client-server-api/#receipts.

`receiptType` is used when the user wants to affect their notification count but
doesn't want other users to see their read receipt.

To move the user's private read receipt to `$123` the client can make a POST
request such as this.

```HTTP
POST /_matrix/client/r0/rooms/!a:example.org/receipt/m.read.private/$123
{}
```

The MSC also proposes adding `m.fully_read` as a possible `receiptType` for `/receipt`.

> To also move the user's `m.fully_read` marker and `m.read` receipt the client
> can make a POST request such as this.

Reviewer 6: ```suggestion
Alternatively, the client can move the user's `m.fully_read` marker and/or 
`m.read` receipt at the same time as `m.read.private` by making a POST
request to the [`/read_markers` endpoint](https://spec.matrix.org/v1.3/client-server-api/#post_matrixclientv3roomsroomidread_markers). For example:
```

```HTTP
POST /_matrix/client/r0/rooms/!a:example.org/read_markers
{
    "m.fully_read": "$123",
    "m.read": "$123",

>     "m.read.private": "$123",

Reviewer 6: ```suggestion
    "m.read.private": "$123"
```

}
```

Both `m.read` and `m.read.private` clear notifications in the same way. If the
user sent two receipts into a room, the later one should be the one that decides
the notification count.

If the user has `m.read` and `m.read.private` receipts at the same event, the
`m.read` receipt takes precedence - an `m.read.private` receipt sent to an event
that already has `m.read`, shouldn't move the user's receipt up in the timeline

> from the perspective of other users.

Reviewer 4: Is there a way for the client to tell, which one is higher, when they come down /sync?

Reviewer 16: Only the latest one is received down sync, not both. The latest one essentially overwrites the earlier one.

Reviewer 4: That applies to one receipt type, but if you have multiple receipt types, do you still only get one of them in /sync? Or do you get both of them and then just need to store both and check which one is later, if you have the full range of events locally?

I guess you can always account for both when evaluating notifications locally, but this added actually quite some complexity that I didn't expect, since I can't simplify it down to one marker. I basically can't go "okay, the marker points to this events, so the event after it has no read marker". I now need to find both and hope they are in the same continuous chunk.

It is not such a big issue, but slightly more annoying than the one receipt, that was hidden or not, which we had before.

Reviewer 2: I am not completely sure I follow - I am pretty sure we send both `m.read` and `m.read.private` down `/sync`. The client then has two receipt types to worry about which adds a bit of complexity when looking for things such as event read-up to - an example of the handling for that is https://github.com/matrix-org/matrix-js-sdk/pull/2221

Reviewer 4: Basically if you have a timeline fragment `A->B->C->D->E`, if your latest `m.read` points at C, you can't know if the later message D is read, unless you know if `m.read.private` points to a previous or later event. I.e. if your fragment currently available to the client is `A->B->C->D`, but `m.read.private` points at `E`, you need to paginate forward until `E` to know for sure, if `D` is read or not. But `m.read.private` might point at `A`, in which case your forward pagination might never give you any knowledge about the private read receipt, since you needed to paginate backwards.

Now in a typical /sync scenario, that isn't too bad, since you have all the newest messages. And usually only then you care about the read receipt.

But some clients implement opening a room at where the user has their read receipt. Now before you could just fetch the context of your latest read receipt. Now you can't do that, since you wouldn't know which fragment comes first for either receipt. Arguably clients should be using the readmarker in this case anyway.

Nonetheless, it would probably be a good idea, if the server could annotate read receipts with some additional information if a read receipt is ordered before or after the other read receipt type. Although now that I think of it, a client can probably just use the timestamp to order read receipts in most cases. I guess it depends on if clients can trust the timestamps to be a reliable way to order read receipts.

Reviewer 16: Synapse has a couple of tests that seem to assert only one will get returned for a particular event:

https://github.com/matrix-org/synapse/blob/c52abc1cfdd9e5480cdb4a03d626fe61cacc6573/tests/rest/client/test_sync.py#L411-L457

Which doesn't really answer the question.

There is logic to handle the ordering of receipts (see `get_receipts_for_user_with_orderings`), but that doesn't seem to get used by `/sync`?

Reviewer 6: Seems like there is some confusion here. It would be good to get this thread resolved.

EW's logic here (https://github.com/matrix-org/matrix-js-sdk/pull/2221/files#diff-1252c1ee1e945dedde8e3e6e96f4b29013a7fc3cd8b83fabc762f2d0add859a8R2356-R2370) seems like it might be good enough? Or do we need the server to annotate RRs with some sort of ordering (if so, what ordering should it use, given it doesn't necessarily have the full event graph, and even if it does there exists only a *partial* ordering between events)?

Reviewer 2: Would this work for you?

```
When determining how far a user has read messages, a client should choose the
event that is located later in the client's local timeline; if the client isn't
aware of the event's location in the client's local timeline, a simple
comparison of timestamps should suffice; if timestamp comparison fails, private
read receipt should take precedence.
```

Reviewer 4: I think it is better to leave the client behaviour unspecified in this case. What you proposed sounds a bit too expensive and I'd rather store all read receipts and check that way. Tbh, I would have preferred if the server sent me what is the latest read receipt, either by bumping the read marker, having only one receipt type or by annotating them. If that is not going to happen, I am fine with just leaving it unspecified as "client needs to figure out what is latest".

> The `m.read` is now optional as sometimes we only want to send `m.read.private`.

Reviewer 6: where is it optional? I think you mean `/read_markers`, but please say so.

run their own homeservers to ensure it behaves correctly.

## Potential issues

Clients which support read receipts would end up rendering the user's receipt as
jumping down when they send a message. This is no different from how IRC and
similarly bridged users are perceived today.

## Alternatives

It has been suggested to use account data to control this on a per-account
basis. While this might have some benefits, it is much less flexible and would
lead to us inventing a way to store per-account settings in account data which
should be handled in a separate MSC.

Previous iterations of this MSC additionally suggested that having an `m.hidden`
flag on existing read receipts could work, however this feels like assigning too
much responsibility to an existing structure.

## Unstable prefix

While this MSC is not considered stable, implementations should use
`org.matrix.msc2285` as a namespace.

|Release         |Development                      |
|----------------|---------------------------------|
|`m.read.private`|`org.matrix.msc2285.read.private`|

To detect server support, clients can either rely on the spec version (when

> stable) or the presence of a `org.matrix.msc2285` flag in  `unstable_features`

Reviewer 4: Could we get a new unstable flag? Otherwise notification clearing just breaks on older servers with no way to tell users why.

Reviewer 4: See https://github.com/Nheko-Reborn/mtxclient/issues/76

Reviewer 18: > Otherwise notification clearing just breaks on older servers with no way to tell users why.

Right next to the experimental flag they should have a note "If your server isn't also running experimental code then this experiment may yield bad results" - unless of course the client is using unstable MSCs to back "stable" features.

Reviewer 4: Preferably a client could also show if a server supports the MSC. For that reasons MSCs usually update the unstable flag, when there is a major change it the MSC.

Reviewer 2: I did consider updating the unstable flag but decided against it since the MSC is rather small and experimental features are sort of expected to break... I wonder if it's still worth it - we could update it but for it to work we would have to wait for one more release cycle, while now if you're up to date everything should work

Reviewer 4: I guess I'll just add a warning then, that we have no way to check for server support, so it might be broken.

Reviewer 2: > I guess I'll just add a warning then, that we have no way to check for server support, so it might be broken.

Sorry for not handling this better

Reviewer 4: No worries, I could have noticed that earlier. And it is an MSC after all :D

Author: Spec-wise we're supposed to update the unstable flag, however for this case I don't think it's needed necessarily. It's a good test to ensure clients aren't overly relying on unstable functionality.

> on `/versions`. Clients are recommended to check for server support to ensure

Reviewer 6: Is there an intention to use `m.read.private` as a "stable" feature before it makes it into a released spec version (and that version being fully adopted by servers)? We might need to consider how clients identify support for it if so.

Reviewer 13: MSC3440 (threads) used `org.matrix.msc3440.stable` as an unstable feature until the spec release, not sure if that's considered standard now

Reviewer 2: From what T said, it sort of is, so that is what we went with

they are not misleading the user about their read receipts not being visible to

> other users.

Reviewer 6: +1; this needs more fleshing out to be reliable.

> other users.

Reviewer 2: I am not sure I completely follow - what is the exact situation here? A client sending `m.read.private` to a server that does not support it? I don't think there is much we can do other than https://github.com/matrix-org/matrix-spec-proposals/pull/2285#discussion_r908466863

# MSC2285: Private read receipts

Currently users must send read receipts in order to affect their notification
counts, which alerts other people that the user has read their message. For
primarily privacy reasons, it may be desirable to users to not advertise to
others that they've read a message.

## Proposal

This MSC proposes adding a new `receiptType` (see [the receipts
spec](https://spec.matrix.org/v1.3/client-server-api/#receipts)) of
`m.read.private`. This `receiptType` is used when the user wants to affect their
notification count but doesn't want other users to see their read receipt.

To move the user's private read receipt to `$123` the client can make a POST
request to the [`/receipt`
endpoint](https://spec.matrix.org/v1.3/client-server-api/#post_matrixclientv3roomsroomidreceiptreceipttypeeventid).
For example:

```HTTP

> POST /_matrix/client/r0/rooms/!a:example.org/receipt/m.read.private/$123

Reviewer 6: These should really be `/v3/` nowadays btw.

While this MSC is not considered stable, implementations should use
`org.matrix.msc2285` as a namespace.

|Stable          |Unstable                         |
|----------------|---------------------------------|
|`m.read.private`|`org.matrix.msc2285.read.private`|

Clients are required to check for server support to ensure they are not sending
read receipts which are not clearing notifications.

While the MSC is considered unstable, to detect server support clients can rely
on the presence of the `org.matrix.msc2285` flag in `unstable_features` on
`/versions`.

> While this MSC is considered stable, to determine server support clients can

Author: ("merged" is a bit confusing when it can mean two different stages of the MSC process: do we mean to say post-FCP?)

rely on the presence of the `org.matrix.msc2285.stable` flag in
`unstable_features` (while the MSC is not in a spec version) or on the presence

> of a suitable spec version in `versions` on `/versions`.

Author: I think this language is from a diagram I provided:
![image](https://user-images.githubusercontent.com/1190097/176955739-7be76c73-41b6-4ae6-b1fe-869078382c0c.png)

"suitable" is my 2am wording for "whatever it was released in".

> of a suitable spec version in `versions` on `/versions`.

Reviewer 6: Ok, speaking to @turt2live:

No, clients are *not* allowed to assume this will be in v1.4. They can check for `org.matrix.msc2285.stable`, but note that server implementations can/should drop that flag at the same time (or soon after) they start advertising the spec version that supports it (probably, but not necessarily, v1.4).

The corollary of that is that a client that want to adopt this feature before it becomes stable must choose between:

 * checking for `org.matrix.msc2285.stable`, and sending public read receipts if that flag is not present, risking read-receipts suddenly becoming public if the server is upgraded to spec vNext before the client.
 * sending private read receipts regardless of whether it sees server-side support for them, and risk notification hell if the server doesn't actually support them.

The first choice might be improved by (a) making sure to accept spec vNext before the server implementation is updated; (b) warning the user if it notices the flag disappearing.

@SimonBrandner: I hope that is clear, if not entirely satisfactory. It would be good to have some words in the MSC reflecting this situation.

> of a suitable spec version in `versions` on `/versions`.

Author: For clarity, the client can also decide that it hates both options and not support private read receipts until it's in a released version of the spec: the safest, but also slowest, option.

While this MSC is not considered stable, implementations should use
`org.matrix.msc2285` as a namespace.

|Stable (post-FCP)|Unstable                         |
|-----------------|---------------------------------|
|`m.read.private` |`org.matrix.msc2285.read.private`|

## Detecting server support

Clients are required to check for server support to ensure they are not sending
read receipts which are not clearing notifications.

> If a client has this feature enabled, it is recommended that, in the case of the

Reviewer 6: recommended? I don't think it has a choice?

server not supporting the MSC, the client should either keep sending private
read receipts and risk notifications not clearing (if the server really does not

> support the MSC) or it should warn the user and start sending public read

Reviewer 6: this is confusing. what do you mean by "the server not supporting the MSC" vs "the server really does not
support the MSC" ?

Reviewer 2: Nothing really :sweat_smile: - it makes more sense without it

Clients are required to check for server support to ensure they are not sending
read receipts which are not clearing notifications.

If a client has this feature enabled, in the case of the server not supporting
the MSC, the client should either keep sending private read receipts and risk
notifications not clearing or it should warn the user and start sending public

> read receipts.

Reviewer 6: sorry, but this still isn't accurate.

If the server doesn't support the MSC and the client keeps sending private RRs, then it is not a *risk* but a *certainty* that notifications will not be cleared.

The difficulty is that a client written today has to rely on the presence of `org.matrix.msc2285` or `org.matrix.msc2285.stable` in `unstable_features`, and that at some point servers will stop advertising that. So it is possible that a server supports the MSC without the client being able to know that. You need to call that out, and mention that,  to avoid this problem, clients need to be updated to check the spec version asap after the release of the spec.

Reviewer 3: If the server doesn't support m.read.private and the client sends just that fields, the server will return an error because m.read was a required field previously.

So a client can just try sending m.private and see if it returns an error
